src/pudl/transform/ferc1.py

```diff
@@ -960,75 +1124,332 @@ def calculate_values_from_components(
+        )
+
+    def groupby_by(self: Self) -> list[str]:
+        """The list of columns to group by.
+
+        We want to default to adding the table_name into all groupby's, but two of our
+        ``by`` options need special treatment.
+        """
+        gb_by = ["table_name", self.by]
+        if self.by in ["ungrouped", "table_name"]:
+            gb_by = [self.by]
+        return gb_by
+
+    def calculate_metric(self: Self, df: pd.DataFrame) -> pd.Series:
+        """Calculate the frequency with which records are tagged as errors."""
+        df["is_not_close"] = self.is_not_close(df)
+        return df.groupby(by=self.groupby_by()).apply(self.metric)
+
+    def snake_case_metric_name(self: Self) -> str:
+        """Convert the TitleCase class name to a snake_case string."""
+        class_name = self.__class__.__name__
+        return re.sub("(?!^)([A-Z]+)", r"_\1", class_name).lower()
+
+    def check(self: Self, calculated_df) -> pd.DataFrame:
+        """Make a df w/ the metric, tolerance and is_error columns."""
+        self.check_required_cols_df(calculated_df)
+        # ungrouped is special because the rest of the stock group names are column
+        # names.
+        if self.by == "ungrouped":
+            calculated_df = calculated_df.assign(
+                ungrouped="ungrouped",
+            )
+
+        metric_name = self.snake_case_metric_name()
+        df = (
+            pd.DataFrame(self.calculate_metric(calculated_df), columns=[metric_name])
+            .assign(
+                **{  # totolerance_ is just for reporting so you can know of off you are
+                    f"tolerance_{metric_name}": self.metric_tolerance,
+                    f"is_error_{metric_name}": lambda x: x[metric_name]
+                    > self.metric_tolerance,
+                }
+            )
+            # make a uniform multi-index w/ group name and group values
+            .assign(group=self.by)
+            .reset_index()
+            .rename(columns={self.by: "group_value"})
+        )
+        # we want to set the index values as the same for all groups, but both the
+        # ungrouped and table_name group require a special exception because we need
+        # to add table_name into the columns
+        if self.by == "ungrouped":
+            df = df.assign(table_name="ungrouped")
+        if self.by == "table_name":
+            # we end up having two columns w/ table_name values in order to keep all the
+            # outputs having the same indexes.
+            df = df.assign(table_name=lambda x: x["group_value"])
+        return df.set_index(["group", "table_name", "group_value"])
+
+
+class ErrorFrequency(GroupMetricCheck):
+    """Check error frequency in XBRL calculations."""
+
+    def metric(self: Self, gb: DataFrameGroupBy) -> pd.Series:
+        """Calculate the frequency with which records are tagged as errors."""
+        try:
+            out = gb[gb.is_not_close].shape[0] / gb.shape[0]
```

Comment:
I used the `shape` attributes here rather than `len()` because it doesn't require any computation to look up the shape, and we're doing this potentially a bazillion times, inside every group.

Reply:
is this just an fyi? or should i add this into the docs/comments?